NotificationServiceExtension/NotificationService.swift

```diff
@@ -2,57 +2,178 @@
         return controller
     }()
+    private let sharedCache = SharedContainerCache<PushNotificationsCache>.init(pathComponent: .pushNotificationsCache, defaultCache: { PushNotificationsCache(settings: .default, notifications: []) })
+    
+    private let fallbackPushContent = WMFLocalizedString("notifications-push-fallback-body-text", value: "New activity on Wikipedia", comment: "Fallback body content of a push notification whose content cannot be determined. Could be either due multiple notifications represented or errors.")
 
     override func didReceive(_ request: UNNotificationRequest, withContentHandler contentHandler: @escaping (UNNotificationContent) -> Void) {
         self.contentHandler = contentHandler
         
-        //TODO: We will modify content in all error cases to "New activity on Wikipedia" (see: https://phabricator.wikimedia.org/T288773 > Push Notification Content section)
         guard let bestAttemptContent = request.content.mutableCopy() as? UNMutableNotificationContent else {
             contentHandler(request.content)
             return
         }
         
+        bestAttemptContent.body = fallbackPushContent
         self.bestAttemptContent = bestAttemptContent
         
-        //TODO: We will modify this to pull the first page of only **unread** notifications, specifying a notnotifiertype=push parameter in the API call. This needs to be done as the next part of PRs for https://phabricator.wikimedia.org/T287310
-        //This temporary call is just to confirm an authenticated call now works from an extension.
-        //remoteNotificationsController.fetchFirstPageNotifications {
-            bestAttemptContent.title = "\(bestAttemptContent.title) [modified]"
-            contentHandler(bestAttemptContent)
-        //}
+        //TODO: Should we consider versioning here? Bail now and show fallback content if current content is anything other than "checkEchoV1".
+        
+        let cache = sharedCache.loadCache()
+        let project = RemoteNotificationsProject.language(cache.settings.primaryLanguageCode, nil, nil)
+        
+        apiController.getUnreadPushNotifications(from: project) { [weak self] newNotifications, error in
+            
+            DispatchQueue.main.async {
+                guard let self = self,
+                      error == nil else {
+                    contentHandler(bestAttemptContent)
+                    return
+                }
+                
+                let finalNotifications = self.determineNewNotificationsAndUpdateCache(newNotifications: newNotifications, cache: cache)
+                
+                //specific handling for talk page types (New messages title, bundled body)
+                let handledAsTalkPage = self.handleTalkPageTypeNotificationsIfNeeded(notifications: finalNotifications, bestAttemptContent: bestAttemptContent, contentHandler: contentHandler)
```

Comment:
@mcleinman Absolutely, I refactored this to your suggestion and it feels so much cleaner, thanks. While I was here I decided to also pull out a lot of this data-heavy business logic code into a separate class called `NotificationServiceHelper` (sidenote: I hate "helper" in names, so I'm happy to rename it if you think of something better). Because this logic is such a pain to manually test, I also wrote unit tests on it. I'm not sure how clear the tests are but it feels better to have something over nothing. I'm in the process of manually going through the PR test steps again to confirm that everything still works.